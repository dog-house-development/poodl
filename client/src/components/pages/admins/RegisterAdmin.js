import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import _ from 'lodash';

import AdminActions from '../../../actions/adminActions';
import DynamicForm from '../../ui/DynamicForm';
import adminInputs from './adminInputs';

const propTypes = {
    auth: PropTypes.object.isRequired,
    errors: PropTypes.object.isRequired
};

export class RegisterAdmin extends Component {
    constructor(props) {
        super(props);
        this.state = {
            seniorCenterId: props.adminSeniorCenterId
        };
    }

    handleChange = e => {
        this.setState({ [e.target.id]: e.target.value });
    };

    onSubmit = () => {
        const admin = {
            ...this.state
        };

        this.props.adminActions.create(admin, this.props.history);
    };

    render() {
        return (
            <div className="register-container page-container">
                <Link to="/dashboard" className="button small tertiary icon">
                    <i className="material-icons button-icon">keyboard_backspace</i> Back to home
                </Link>
                <div className="panel">
                    <h1 className="panel-title">Register Admin</h1>
                    <DynamicForm
                        inputs={adminInputs}
                        onChange={this.handleChange}
                        onSubmit={this.onSubmit}
                        submitButtonLabel="Register Admin"
                        errors={this.props.errors}
                        values={this.state}
                        loading={this.props.loading}
                    />
                </div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        auth: state.auth,
        adminIsSuper: _.get(state.auth.admin, 'accessLevel', false) === 'Super',
        adminSeniorCenterId: _.get(state.auth.admin, 'seniorCenterId'),
        errors: state.admins.errors,
        loading: state.admins.loading
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        adminActions: bindActionCreators(AdminActions, dispatch)
    };
};

RegisterAdmin.propTypes = propTypes;
export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(RegisterAdmin));
