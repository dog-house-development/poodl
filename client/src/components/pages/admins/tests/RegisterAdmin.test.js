import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import { RegisterAdmin, mapStateToProps, mapDispatchToProps } from '../RegisterAdmin';

configure({ adapter: new Adapter() });

describe('Register tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                auth: {
                    isAuthenticated: true,
                    loading: false,
                    admin: {
                        accessLevel: 'Admin',
                        seniorCenterId: 123,
                        firstName: 'Frog',
                        lastName: 'Anderson'
                    },
                    errors: {}
                },
                admins: {
                    errors: {}
                }
            },
            _state
        );
        props = _.assign({}, { history: ['/admins/register'] }, _props);
        wrapper = shallow(
            <RegisterAdmin
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
        wrapper.setState({
            name: 'Sam',
            email: 'sam@test.com',
            password: 'abc123',
            password2: 'abc123',
            errors: { email: 'Invalid email' }
        });
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                adminIsSuper: false,
                adminSeniorCenterId: 123,
                auth: {
                    isAuthenticated: true,
                    loading: false,
                    admin: {
                        accessLevel: 'Admin',
                        seniorCenterId: 123,
                        firstName: 'Frog',
                        lastName: 'Anderson'
                    },
                    errors: {}
                },
                errors: {}
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(JSON.stringify({ adminActions: {} }));
        });
    });

    describe('onChange', () => {
        it('should run without errors', () => {
            spyOn(instance, 'onChange');
            expect(instance.state.email).toEqual('sam@test.com');
            const e = {
                target: { id: 'email', value: 'Meow' }
            };
            instance.onChange(e);
        });
    });

    describe('onSubmit', () => {
        xit('should run without errors', () => {
            spyOn(instance, 'onSubmit');
            const e = {
                target: { id: 'email', value: 'Meow' },
                preventDefault: () => {}
            };
            wrapper.find('Form').simulate('submit', e);
        });
    });

    describe('getFields', () => {
        xit('should return correct fields if not super admin', () => {
            expect(instance.getFields()).toEqual([
                {
                    onChange: instance.onChange,
                    id: 'firstName',
                    type: 'text',
                    label: 'First name',
                    placeholder: 'John...',
                    sidebyside: 1,
                    autoComplete: 'off',
                    error: undefined
                },
                {
                    onChange: instance.onChange,
                    id: 'lastName',
                    type: 'text',
                    label: 'Last name',
                    placeholder: 'Smith...',
                    sidebyside: 2,
                    autoComplete: 'off',
                    error: undefined
                },
                {
                    onChange: instance.onChange,
                    id: 'email',
                    type: 'email',
                    label: 'Email',
                    placeholder: 'example@poodl.com...',
                    autoComplete: 'off',
                    error: 'Invalid email'
                },
                {
                    onChange: instance.onChange,
                    id: 'password',
                    type: 'password',
                    label: 'Password',
                    placeholder: 'Shhhhh...',
                    autoComplete: 'off',
                    error: undefined
                },
                {
                    onChange: instance.onChange,
                    id: 'password2',
                    type: 'password',
                    label: 'Confirm Password',
                    placeholder: 'Again...',
                    autoComplete: 'off',
                    error: undefined
                }
            ]);
        });

        xit('should return correct fields if super admin', () => {
            setInstanceAndWrapper({}, { auth: { admin: { accessLevel: 'Super', seniorCenterId: '123' } } });
            expect(instance.getFields()).toEqual([
                {
                    onChange: instance.onChange,
                    id: 'firstName',
                    type: 'text',
                    label: 'First name',
                    placeholder: 'John...',
                    sidebyside: 1,
                    autoComplete: 'off',
                    error: undefined
                },
                {
                    onChange: instance.onChange,
                    id: 'lastName',
                    type: 'text',
                    label: 'Last name',
                    placeholder: 'Smith...',
                    sidebyside: 2,
                    autoComplete: 'off',
                    error: undefined
                },
                {
                    onChange: instance.onChange,
                    id: 'email',
                    type: 'email',
                    label: 'Email',
                    placeholder: 'example@poodl.com...',
                    autoComplete: 'off',
                    error: 'Invalid email'
                },
                {
                    onChange: instance.onChange,
                    id: 'password',
                    type: 'password',
                    label: 'Password',
                    placeholder: 'Shhhhh...',
                    autoComplete: 'off',
                    error: undefined
                },
                {
                    onChange: instance.onChange,
                    id: 'password2',
                    type: 'password',
                    label: 'Confirm Password',
                    placeholder: 'Again...',
                    autoComplete: 'off',
                    error: undefined
                },
                {
                    onChange: instance.onChange,
                    id: 'seniorCenterId',
                    content: '123',
                    type: 'text',
                    label: 'Senior Center Id',
                    placeholder: 'ID...',
                    error: undefined
                }
            ]);
        });
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
