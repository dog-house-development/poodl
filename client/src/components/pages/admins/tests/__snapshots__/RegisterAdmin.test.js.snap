// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register tests onChange should run without errors 1`] = `
<div
  className="register-container page-container"
>
  <Link
    className="button small tertiary"
    replace={false}
    to="/dashboard"
  >
    <i
      className="material-icons"
    >
      keyboard_backspace
    </i>
     Back to home
  </Link>
  <Form
    buttonLabel="Register"
    fields={
      Array [
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "firstName",
          "label": "First name",
          "onChange": [Function],
          "placeholder": "John...",
          "sidebyside": 1,
          "type": "text",
          "value": "Sam",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "lastName",
          "label": "Last name",
          "onChange": [Function],
          "placeholder": "Smith...",
          "sidebyside": 2,
          "type": "text",
          "value": "Smith",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "email",
          "label": "Email",
          "onChange": [Function],
          "placeholder": "example@poodl.com...",
          "type": "email",
          "value": "sam@test.com",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "password",
          "label": "Password",
          "onChange": [Function],
          "placeholder": "Shhhhh...",
          "type": "password",
          "value": "abc123",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "password2",
          "label": "Confirm Password",
          "onChange": [Function],
          "placeholder": "Again...",
          "type": "password",
          "value": "abc123",
        },
      ]
    }
    formTitle="Register Admin"
    noValidate={true}
    onSubmit={[Function]}
  />
</div>
`;

exports[`Register tests render should render correctly 1`] = `
<div
  className="register-container page-container"
>
  <Link
    className="button small tertiary"
    replace={false}
    to="/dashboard"
  >
    <i
      className="material-icons"
    >
      keyboard_backspace
    </i>
     Back to home
  </Link>
  <Form
    buttonLabel="Register"
    fields={
      Array [
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "firstName",
          "label": "First name",
          "onChange": [Function],
          "placeholder": "John...",
          "sidebyside": 1,
          "type": "text",
          "value": "Sam",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "lastName",
          "label": "Last name",
          "onChange": [Function],
          "placeholder": "Smith...",
          "sidebyside": 2,
          "type": "text",
          "value": "Smith",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "email",
          "label": "Email",
          "onChange": [Function],
          "placeholder": "example@poodl.com...",
          "type": "email",
          "value": "sam@test.com",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "password",
          "label": "Password",
          "onChange": [Function],
          "placeholder": "Shhhhh...",
          "type": "password",
          "value": "abc123",
        },
        Object {
          "autoComplete": "off",
          "error": undefined,
          "id": "password2",
          "label": "Confirm Password",
          "onChange": [Function],
          "placeholder": "Again...",
          "type": "password",
          "value": "abc123",
        },
      ]
    }
    formTitle="Register Admin"
    noValidate={true}
    onSubmit={[Function]}
  />
</div>
`;
